### PR TITLE
Adding new features to ci-signal-report

### LINKED ISSUE #2360: ci-signal-reporter, JSON formatted print out
#### JSON formatted print out

:green_circle: ci-signal-report-cli

**Summary:** Report printout and data structs allows json format

<details>
<summary>Current situation</summary>
Currently the report only supports plain text print outs and does not define JSON in data structs (needed: `json:"title"`).
</details>

<details>
<summary>Desired outcome</summary>
It should be possible to print the ci signal report in json format. This enables the pkg to get used in a bunch of more scenarios as of right now. (like prometheus could target a ci-signal-report exporter)
</details>
<br>

**Benefits:** Versatile enhancement / enabler of more use cases  

**Concerns:** None

### LINKED ISSUE #2361: ci-signal-reporter, Specify reporter kind
#### Specify reporter kind

:green_circle: ci-signal-report-cli

**Summary:** Select a reporter to generate only a testgrid or github report

<details>
<summary>Current situation</summary>
Currently a github report & testgrid report is getting generated. This leads to printing a buch of stuff if only the testgrid summary is needed (& vise versa)
</details>

<details>
<summary>Desired outcome</summary>
Via a flag it should be possible to specify which reporter should be used to generate the ci-signal-report. 
</details>
<br>

**Benefits:** :: only report what you are looking for, right tool for the job

**Concerns:** None

### LINKED ISSUE #2362: ci-signal-reporter, Detailed testgrid report
#### Detailed testgrid report

:green_circle: ci-signal-report-cli

**Summary:** Give detailed information about current testgrid jobs

<details>
<summary>Current situation</summary>
Currently only a testgrid summary is getting provided which does show testgrid job insights.
</details>

<details>
<summary>Desired outcome</summary>
Similar to the [updated ci signal report](https://github.com/leonardpahlke/ci-signal-report) more details would be listed printed out (without the short flag set)

example:

FAILING 🔴🔴🔴 gce-cos-master-serial
- https://testgrid.k8s.io/sig-release-master-informing#gce-cos-master-serial
- Sig's involved [sig-storage]
- Currently 6 test are failing
- 0 of 9 passed recently

</details>
<br>

**Benefits of the change:** improved testgrid visibility and significance

**Concerns:** None

### LINKED ISSUE #2363: ci-signal-reporter, Generic report template
#### Generic report template

* :green_circle: ci-signal-report-cli

**Summary:** Use a generic report template

<details>
<summary>Current situation</summary>
Currently the testgrid and github report is customized. Its not abstracted. 
</details>

<details>
<summary>Desired outcome</summary>
A generic template should be defined which can be used to print github and testgrid reports.
</details>
<br>

**Benefits:** Using a generic template simplifies the process of making changes to the code (Seperation of concerncs). It also decouples the code which does not allow github reports be printed with testgrrid printers.  

**Concerns:** None

### LINKED ISSUE #2364: ci-signal-reporter, Tablewriter report printer
#### Tablewriter report printer

* :green_circle: ci-signal-report-cli

**Summary:** Print reports in a ascii table [tablewriter](https://github.com/olekukonko/tablewriter)

##### Current situation
Currently the reports are being printed as a list

##### Desired outcome
Printing the report in a table would structure the report and improve visibility. 

**Benefits:** Simpler way to display data  

**Concerns:** None

### PR DESCRIPTION
**Pull request by leonardpahlke**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR implements a couple of planned features to the ci signal report CLI (see referenced issues).

#### Which issue(s) this PR fixes:

* closes #2364
* closes #2363
* closes #2361
* closes #2360
* closes #2362

#### Special notes for your reviewer:

Some things I am thinking about:
* CIReporter interface (root.go): remove method GetCIReporterHead // not sure tho
* Adding colors to testgrid
* Reduce testgrid table width somehow (maybe force line breaks)

#### Does this PR introduce a user-facing change?

```release-note
Add the 'testgrid' command to the CLI, which creates only a testgrid report.
Add the 'github' command to the CLI, which creates only a github report
Add flag '--json' to print the report in json format.
Report will be printed in table format
```

/label tide/merge-method-squash
/assign @puerco

### CODE REVIEW

**Comment by leonardpahlke** [comment]
This is how the report output currently looks like
2021-12-16-ci-report-output

**Comment by leonardpahlke** [comment]
i recently installed a new distro on my main machine and generated a new gpg key (if you are wondering why its says 'Unverified')

**Review by puerco — COMMENTED** [review]
OK, here are some initial comments. Sorry for the delay @leonardpahlke

File: cmd/ci-reporter/cmd/root.go [hunk 1]
```diff
@@ -42,57 +46,186 @@ func Execute() error {
 	return rootCmd.Execute()
 }
 
-// ReporterConfig configuration that is getting injected into ci-signal report functions
-type ReporterConfig struct {
+// Config configuration that is getting injected into ci-signal report functions
+type Config struct {
 	GithubClient   *github.Client
```
[the remark this answers is not in the document]
Comment (leonardpahlke) [reply cmd/ci-reporter/cmd/root.go:51]: Yes! I would do that after features have been added to the tool (see [project board](https://github.com/orgs/kubernetes/projects/65/))

**Comment by leonardpahlke** [comment]
No problem! Thanks for the review. I updated the code.

**Review by puerco — APPROVED** [review]
Looking good, thanks @leonardpahlke!

**Pull request merged**